runtime/precompiles/src/staking.rs

```diff
@@ -0,0 +1,348 @@
+{
+	// In solidity all values are encoded to this width
+	const AMOUNT_SIZE_BYTES: usize = 32;
+
+	if input.len() != AMOUNT_SIZE_BYTES {
+		log::info!(
+			"Unable to parse amount. Got {} bytes, expected {}",
+			input.len(),
+			AMOUNT_SIZE_BYTES,
+		);
+		return Err(ExitError::Other(
+			"Incorrect input length for amount parsing".into(),
+		));
+	}
+
+	let amount: Balance = U256::from_big_endian(&input[0..AMOUNT_SIZE_BYTES])
+		.try_into()
+		.map_err(|_| ExitError::Other("Amount is too large for provided balance type".into()))?;
+	Ok(amount)
+}
+
+impl<Runtime> ParachainStakingWrapper<Runtime>
+where
+	Runtime: parachain_staking::Config + pallet_evm::Config,
+	Runtime::AccountId: From<H160>,
+	BalanceOf<Runtime>: TryFrom<U256> + Debug,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	Runtime::Call: From<parachain_staking::Call<Runtime>>,
+{
+	// The accessors are first. They directly return their result.
+
+	fn is_nominator(input: &[u8]) -> Result<(ExitSucceed, Vec<u8>, u64), ExitError> {
+		// parse the address
+		let nominator = H160::from_slice(&input[12..32]);
+
+		log::info!("Checking whether {:?} is a nominator", nominator);
+
+		// fetch data from pallet
+		let is_nominator = parachain_staking::Pallet::<Runtime>::is_nominator(&nominator.into());
+
+		log::info!("Result from pallet is {:?}", is_nominator);
+
+		// Solidity's bool type is 256 bits as shown by these examples
+		// https://docs.soliditylang.org/en/v0.8.0/abi-spec.html
+		let mut result_bytes = [0u8; 32];
+		if is_nominator {
+			result_bytes[31] = 1;
+		}
+
+		log::info!("Result bytes are {:?}", result_bytes);
+
+		//TODO figure out how much gas it costs to check whether you're a nominator.
+		// That function will not naturally be benchmarked because it is not dispatchable
+		// I guess the heavy part will be one storage read.
+		let gas_consumed = 0;
+
+		return Ok((ExitSucceed::Returned, result_bytes.to_vec(), gas_consumed));
+	}
+
+	fn is_candidate(input: &[u8]) -> Result<(ExitSucceed, Vec<u8>, u64), ExitError> {
+		// parse the address
+		let candidate = H160::from_slice(&input[12..32]);
+
+		log::info!("Checking whether {:?} is a collator candidate", candidate);
+
+		// fetch data from pallet
+		let is_candidate = parachain_staking::Pallet::<Runtime>::is_candidate(&candidate.into());
+
+		log::info!("Result from pallet is {:?}", is_candidate);
+
+		// Solidity's bool type is 256 bits as shown by these examples
+		// https://docs.soliditylang.org/en/v0.8.0/abi-spec.html
+		let mut result_bytes = [0u8; 32];
+		if is_candidate {
+			result_bytes[31] = 1;
+		}
+
+		log::info!("Result bytes are {:?}", result_bytes);
+
+		//TODO figure out how much gas it costs to check whether you're a collator candidate.
+		// That function will not naturally be benchmarked because it is not dispatchable
+		// I guess the heavy part will be one storage read.
```

Comment:
It could be benchmarked -- benchmarking does support non-dispatchables. If we use that strategy to calculate gas, we'll also want to respect weight -> gas mapping here.

Reply:
We would only need benchmarking for weight. Then we'll use the weight-gas-mapping at runtime.